test/studies/shootout/reverse-complement/bharshbarg/revcomp-buf.chpl

```diff
@@ -0,0 +1,155 @@
+/* The Computer Language Benchmarks Game
+   http://benchmarksgame.alioth.debian.org/
+
+   contributed by Ben Harshbarger
+   derived from the Rust #2 version by Matt Brubeck
+*/
+
+/*
+   This is very ugly because we don't have good IO support for
+   reading/writing buffers. The 'data' buffer is included in the
+   'buf' record so that we know exactly how many bytes to read
+   with 'readBytes'. This is so that we don't accidentally read
+   past the end of the file and get an error. This would be bad
+   because 'readBytes' doesn't tell us how many bytes it actually
+   read...
+*/
+
+config param columns = 61;
+
+const table = initTable("ATCGGCTAUAMKRYWWSSYRKMVBHDDHBVNN\n\n");
+
+record buf {
+  const bufSize : int;
+  var buf : [0..#bufSize] uint(8);
+  var cur, cap, numLeft : int;
+  var chan : channel(writing=false, kind=iokind.native, locking=false);
+
+  proc init(fi:file, bs:int) {
+    this.bufSize = bs;
+
+    super.init();
+
+    chan = fi.reader(locking=false);
+    numLeft = fi.length();
+  }
+
+  pragma "no copy return"
+  proc fill() {
+    if cur >= cap {
+      if numLeft > 0 {
+        cap = min(bufSize, numLeft);
+        chan.readBytes(c_ptrTo(buf), cap);
+        numLeft -= cap;
+
+        // ensure we return an empty slice if we run out of bytes
+        cur = 0;
+      } else {
+        cur = 0;
+        cap = 0;
+      }
+    }
+    const low = if cap == 0 then 1 else cur;
+    return buf[low..max(0, cap-1)];
+  }
+
+  proc consume(n : int) {
+    cur = min(cur + n, cap);
+  }
+
+  proc _memchr(c : uint(8), arr : []) {
+    extern proc memchr(s:c_void_ptr, c : c_int, n : size_t) : c_void_ptr;
+    const ptr = c_ptrTo(arr);
+    const ret = memchr(ptr, c:c_int, arr.size:size_t);
+    if ret != c_nil {
+      const idx = arr.domain.first + ret:int - ptr:int;
+      return idx;
+    }
+    return -1;
+  }
+
+  proc readUntil(term : uint(8), data : [] uint(8)) : int {
+    var read = 0;
+    while true {
+      var done = false, used = 0;
+      ref avail = fill();
+      if avail.size > 0 {
+        const idx = _memchr(term, avail);
+        if idx >= 0 {
+          data.push_back(avail[..idx]);
```

Comment:
For a while I was confused about this line. Could you add a comment here or at the function level, to indicate that it appends a big block to the array once it's available?